File: Swift/TuringMachine.swift
class TuringMachine {
    var tape: [Character]
    var head: Int
    var state: String

    init(input: String) {
        self.tape = Array("  \(input)  ") // Tape with input string and extra spaces
        self.head = 2 // Start at the first character of the input
        self.state = "q0" // Initial state
    }

    func execute() {
        while state != "accept" && state != "reject" {
            let currentSymbol = tape[head]

            switch state {
            case "q0":
                if currentSymbol == "a" {
                    tape[head] = "X"
                    head += 1
                    state = "q1"
                } else if currentSymbol == " " {
                    state = "accept"
                } else {
                    state = "reject"
                }

            case "q1":
                if currentSymbol == "a" || currentSymbol == "X" {
                    head += 1
                } else if currentSymbol == "b" {
                    tape[head] = "Y"
                    head += 1
                    state = "q2"
                } else {
                    state = "reject"
                }

            case "q2":
                if currentSymbol == "b" || currentSymbol == "Y" {
                    head += 1
                } else if currentSymbol == " " {
                    state = "accept"
                } else {
                    state = "reject"
                }

            default:
                state = "reject"
            }
        }

        if state == "accept" {
            print("Input is accepted by the Turing Machine.")
        } else {
            print("Input is rejected by the Turing Machine.")
        }
    }
}

let input = "aabb"
let tm = TuringMachine(input: input)
tm.execute()

File: Ruby/TuringMachine.rb
class TuringMachine
  def initialize(input)
    @tape = "  #{input}  ".chars # Tape with input string and extra spaces
    @head = 2 # Start at the first character of the input
    @state = "q0" # Initial state
  end

  def execute
    while @state != "accept" && @state != "reject"
      current_symbol = @tape[@head]

      case @state
      when "q0"
        if current_symbol == "a"
          @tape[@head] = "X"
          @head += 1
          @state = "q1"
        elsif current_symbol == " "
          @state = "accept"
        else
          @state = "reject"
        end

      when "q1"
        if ["a", "X"].include?(current_symbol)
          @head += 1
        elsif current_symbol == "b"
          @tape[@head] = "Y"
          @head += 1
          @state = "q2"
        else
          @state = "reject"
        end

      when "q2"
        if ["b", "Y"].include?(current_symbol)
          @head += 1
        elsif current_symbol == " "
          @state = "accept"
        else
          @state = "reject"
        end
      end
    end

    if @state == "accept"
      puts "Input is accepted by the Turing Machine."
    else
      puts "Input is rejected by the Turing Machine."
    end
  end
end

input = "aabb"
tm = TuringMachine.new(input)
tm.execute
